ruma-serde/src/canonical_json/value.rs

```diff
@@ -126,20 +161,19 @@ impl fmt::Display for CanonicalJsonValue {
     /// assert_eq!(compact,
     ///     "{\"city\":\"London\",\"street\":\"10 Downing Street\"}");
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", serde_json::to_string(&self).map_err(|_| fmt::Error)?)
+        write!(f, "{}", to_json_string(&self).map_err(|_| fmt::Error)?)
     }
 }
 
 impl TryFrom<JsonValue> for CanonicalJsonValue {
-    type Error = serde_json::Error;
+    type Error = CanonicalError;
+
     fn try_from(json: JsonValue) -> Result<Self, Self::Error> {
         Ok(match json {
             JsonValue::Bool(b) => Self::Bool(b),
             JsonValue::Number(num) => Self::Integer(
-                Int::try_from(num.as_i64().ok_or_else(|| {
-                    serde_json::Error::custom("Invalid number found, expected i64")
-                })?)
-                .map_err(serde_json::Error::custom)?,
+                Int::try_from(num.as_i64().ok_or(CanonicalError::IntDecimal)?)
```

Comment:
`as_i64` can also fail if the input is larger than the max value for i64, but smaller or equal to `u64::MAX`. Also I think serde has some form of i128 / u128 support. So I'm pretty sure `as_i64` can also fail without any decimal places being present.

Reply:
Now there is one variant called `IntConvert` for all number conversion errors.